### ISSUE TITLE
Refactored Yii2 codeception module breaks the app lifecycle

### ISSUE THREAD

**Issue by russianlagman** [Codeception/Codeception#4910]
#### What are you trying to achieve?

Currently I'm bootstraping the following code in my api modules:

```php
Yii::$app->response->on(Response::EVENT_BEFORE_SEND, function ($event) {
            if (Yii::$app->request->url == '/rest' || Yii::$app->request->url == '/rest/') {
                return;
            }

            /** @var Response $response */
            $response = $event->sender;
            $response->format = Response::FORMAT_JSON;
            if ($response->statusCode < 400) {
                $response->data = [
                    'success' => true,
                    'status' => $response->statusCode,
                    'data' => $response->data,
                ];
            } else {
                ArrayHelper::remove($response->data, 'status');
                ArrayHelper::remove($response->data, 'type');
                $response->data = [
                    'success' => false,
                    'status' => $response->statusCode,
                    'data' => $response->data,
                ];
            }
        });
```

The following test does not work anymore:

```
        $I->seeResponseCodeIs(200);
        $I->seeResponseIsJson();
        $I->seeResponseContainsJson(['success' => true]);
```
#### What do you get instead?

Response object looses it's event listeners. 

* Codeception version: 2.4.1
* PHP Version: 7.1
* Operating System: Linux
* Installation type: Composer

**Comment by SamMousa** [comment]
Some of the discussion happened in the PR.
Todo:
1. Events should obviously be supported
2. A new test case should be added to the `codeception/yii2-tests` repo.
3. We should think on other changes besides events that need support.

**Comment by russianlagman** [comment]
I think that app instance persitance should be optional and disabled by default. I have seen a lot of controllers changing the app components state. Like `Yii::$app->log->targets = []`

**Comment by SamMousa** [comment]
The app is persisted during the test, there are just 2 exception, the request and the response object...

**Comment by SamMousa** [comment]
One issue with these things, is contamination.
Consider the above code while using the test below.
```php
public function testContamination(FunctionalTester $I)
{
	$I->amOnPage('/rest');
	$I->assertEquals('json', \Yii::$app->response->format);
	// This code does not set any response format
	$I->amOnPage('/not-rest');
        /* This will fail since the previous REST request is actually changing server state. */
	$I->assertEquals('html', \Yii::$app->response->format); 
}
```

There is no solution that will resolve all cases while fullfilling all requirements.
1. If we want to have the response object available for inspection we cannot recreate the whole application after a request.
2. If we want to use database transaction we cannot recreate the whole application after a request.
3. If we want to be able to inspect the response object after a request we must delay recreating it until we do the next request.
4. If we don't want contamination we must recreate the response object before each request.
5. If we want to be able to alter application state before a request we cannot recreate the application before each requests.

In conclusion, these event handlers alter application state and there's no way to automatically decide how that will or should affect subsequent requests. An event attached in a controller is not expected to fire for routes that go to a different controller.

One solution proposed by @russianlagman is to show warnings when doing multiple requests inside the same tests. I feel that having multiple requests is a common thing in tests so that solution would severely limit usability.

**Comment by askobara** [comment]
Hello there.
Looks like after the PR was merged, global events can't be tested because of [this](https://github.com/Codeception/Codeception/pull/4894/files?diff=split#diff-94abe3ee46e86aa1477f9565b6d2f9f5R43).
Of course, it's could be my fault, and I should avoid creating events in files such `common/config/bootstrap.php` files, but it worked recently.

**Comment by SamMousa** [comment]
@askobara can you give me a code sample?

**Comment by askobara** [comment]
I hope this would be enough.

```php
// common/config/bootstrap.php

use yii\base\Event;
use yii\db\ActiveRecord;
use common\models\Model;

Event::on(Model::class, ActiveRecord::EVENT_BEFORE_INSERT, function ($event) {
    $model = $event->sender;
    if ($model->value == 1) {
        $model->another_value = 100;
    }
});
```
```php
// common/models/Model.php

namespace common\models;

use Yii;
use yii\db\ActiveRecord;

/**
 * {@inheritdoc}
 */
class Model extends ActiveRecord
{
    /**
     * @inheritdoc
     */
    public static function tableName()
    {
        return '{{%models}}';
    }

    /**
     * @inheritdoc
     */
    public function rules()
    {
        return [
            ['value', 'boolean'],
            ['another_value', 'integer'],
        ];
}

```

```php
// common/tests/unit

use common\models\Model;

class ModelTest extends \Codeception\Test\Unit
{
    use \Codeception\Specify;

    /**
     * @test
     * NOTE: all events will be lost when this one have been tested
     */
    public function tableName()
    {
        expect('tableName is equal to {{%models}}', Model::tableName())->equals('{{%models}}');
    }

    /**
     * @test
     */
    public function events()
    {
        $model = new \common\models\Model();
        $model->value = true;
        $model->save();
        expect('another_value is equal to 100', $model->another_value)->equals(100);
    }
}
```
